processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/TableContainerGenerator.java

```diff
@@ -201,10 +203,58 @@ public TypeSpec generateGtfsContainerClass() {
     typeSpec.addMethod(generateSetupIndicesMethod());
     typeSpec.addMethod(generateForHeaderAndEntitiesMethod());
     typeSpec.addMethod(generateForEntitiesMethod());
+    typeSpec.addMethod(generateGetKeyColumnNames());
+    typeSpec.addMethod(generateByPrimaryKey());
 
     return typeSpec.build();
   }
 
+  private MethodSpec generateGetKeyColumnNames() {
+    return MethodSpec.methodBuilder("getKeyColumnNames")
+        .addAnnotation(Override.class)
+        .addModifiers(Modifier.PUBLIC)
+        .returns(ParameterizedTypeName.get(ImmutableList.class, String.class))
+        .addStatement("return $T.KEY_COLUMN_NAMES", classNames.tableLoaderTypeName())
+        .build();
+  }
+
+  private MethodSpec generateByPrimaryKey() {
+    MethodSpec.Builder method =
+        MethodSpec.methodBuilder("byPrimaryKey")
+            .addAnnotation(Override.class)
+            .addModifiers(Modifier.PUBLIC)
+            .returns(
+                ParameterizedTypeName.get(
+                    ClassName.get(Optional.class), classNames.entityImplementationTypeName()))
+            .addParameter(String[].class, "keys");
+    if (fileDescriptor.primaryKey().isPresent()) {
+      method
+          .beginControlFlow("if (keys.length != 1)")
+          .addStatement("return Optional.empty()")
+          .endControlFlow();
+      method.addStatement(
+          "return Optional.ofNullable($L.getOrDefault(keys[0], null))",
+          byKeyMapName(fileDescriptor.primaryKey().get().name()));
+    } else if (fileDescriptor.sequenceKey().isPresent() && fileDescriptor.firstKey().isPresent()) {
+      method
+          .beginControlFlow("if (keys.length != 2)")
+          .addStatement("return Optional.empty()")
+          .endControlFlow();
+      // TODO(aababilov): Implement this.
```

Comment:
Is it still something to do here?

Reply:
Yes, we need to implement this. One way to do it is to introduce a hash map<composite_key, entity>. The existing containers for stop_times and shapes do not support quick lookup by composite key - they just allow to retrieve a list of all entities that have the same first key (e.g., trip_id).